src/storage/src/source/healthcheck.rs

```diff
@@ -203,15 +204,41 @@ impl Healthchecker {
     ) {
         let since = read_handle.since().clone();
         trace!("Bootstrapping state as of {:?}!", since);
-        // Ensure the collection is readable at `since`
-        if PartialOrder::less_than(&since, &self.upper) {
-            let updates = read_handle
-                .snapshot_and_fetch(since.clone())
-                .await
-                .expect("local since is not beyond read handle's since");
-            self.process_collection_updates(updates);
-        };
-        self.sync(upper).await;
+
+        let mut safe_upper: Antichain<_> =
+            since.elements().iter().map(|t| t.step_forward()).collect();
+        safe_upper.join_assign(upper);
+        if PartialOrder::less_than(upper, &safe_upper) {
+            // It's important that the healthchecker reads the existing state of the collection;
+            // otherwise, we may not be able to initialize the healthchecker with the right status.
+            // However, the collection might not always be immediately readable at all, if `upper <= since`.
+            // Let's advance our upper, and assume that other writers will eventually also do so
+            // so we can eventually make progress.
+            loop {
+                let no_updates: &[((SourceData, ()), Timestamp, i64)] = &[];
+                let advance_result = self
+                    .write_handle
+                    .compare_and_append(no_updates, upper.clone(), safe_upper.clone())
+                    .await;
+
+                match advance_result {
+                    Ok(Ok(Ok(_))) => break,
+                    Ok(Ok(Err(_upper))) => break, // someone else is advancing the upper
```

Comment:
good question! in practise this mostly arises when since == upper, and in that case any progress in the upper is enough to unblock the snapshot. if upper < since, the snapshot may still be blocked after this `break`. in practise a) this should not happen for this particular collection, and b) anyways environmentd should be constantly bumping the upper and get us unstuck sooner or later.

but in retrospect that's lazy and i should go do the more obviously correct / less context-dependent thing here. i'll make that change iff it looks like we're going to be able to reuse this healthcheck!